tests/unit/defrag.cpp

```diff
@@ -0,0 +1,522 @@
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE DEFRAG library test
+
+#include <boost/test/unit_test.hpp>
+#include <boost/filesystem.hpp>
+
+#include <future>
+
+#include "eblob_wrapper.h"
+
+#include "library/blob.h"
+#include "library/datasort.h"
+#include "library/list.h"
+
+#include "eblob/eblob.hpp"
+
+#include <vector>
+#include <utility>
+#include <random>
+#include <iostream>
+#include <functional>
+
+#include <sysexits.h>
+#include <err.h>
+
+
+void init_datasort_cfg(eblob_wrapper &wrapper, datasort_cfg &cfg) {
+	memset(&cfg, 0, sizeof(cfg));
+	cfg.b = wrapper.get();
+	cfg.log = cfg.b->cfg.log;
+}
+
+
+std::vector<char> generate_random_data(std::mt19937 &gen, std::uniform_int_distribution<unsigned> &dist) {
+	size_t datasize = 2 * (1 << 20); // 1Mib
+	if (dist(gen) % 10 != 0) {
+		datasize = 1 + dist(gen) % (1000); // <= 1KiB
+	}
+
+	std::vector<char> data(datasize);
+	for (auto &elem : data) {
+		elem = dist(gen) % 26 + 'a';
+	}
+
+	return data;
+}
+
+
+class iterator_private {
+public:
+	explicit iterator_private(std::vector<Item> &items_)
+	: items(items_)
+	{
+	}
+
+	std::vector<Item> &items;
+	size_t number_checked = 0;
+};
+
+
+int iterate_callback(struct eblob_disk_control *dc,
+		     struct eblob_ram_control *rctl __attribute_unused__,
+		     int fd,
+		     uint64_t data_offset,
+		     void *priv,
+		     void *thread_priv __attribute_unused__)
+{
+	int ret = 0;
+	if (dc->flags & BLOB_DISK_CTL_REMOVE)
+		err(EX_SOFTWARE, "removed dc occured");
+
+	iterator_private &ipriv = *static_cast<iterator_private*>(priv);
+	auto &items = ipriv.items;
+	if (ipriv.number_checked >= items.size())
+		err(EX_SOFTWARE, "index out of range. index: %zd, size: %zd", ipriv.number_checked, items.size());
+
+	auto &item = items[ipriv.number_checked];
```

Comment:
After defrag, are the items iterated in the order they were written?

Reply:
Not necessary. [Here](https://github.com/s-mx/eblob/blob/master/tests/unit/defrag.cpp#L165) we pass a vector which will be compared with iterated data. And there earlier you can see that we sort only part of shadow_elems because in that test only first base is sorted.